### PR TITLE
Input Masking (Meta #69)

### PR DESCRIPTION
**Pull request by mawiswiss**
Used with https://github.com/araise-dev/CrudBundle/pull/76

### CODE REVIEW

File: src/Resources/assets/controllers/input-mask_controller.js [hunk 1]
```diff
@@ -0,0 +1,27 @@
+import { Controller } from "@hotwired/stimulus"
+import IMask from "imask"
```
Comment (renestalder) [inline src/Resources/assets/controllers/input-mask_controller.js]: Consider importing these asynchronously and not synchronously to ensure the land in a separate JavaScript chunk and are only loaded when the controller is connected.

Unless Symfony is doing automatic asynchronous loading of controllers in the background (which I doubt), this is blowing up the JavaScript bundle unnecessarily. A good chunk is under 500 KB.

File: src/Resources/assets/controllers/input-mask_controller.js [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        mask: String,
+        scale: { type: Number, default: 2 },
+        radix: { type: String, default: '.' },
+        thousandsSeparator: { type: String, default: '\'' },
+        normalizeZeros: { type: Boolean, default: false },
+    };
+    connect() {
+        const typeMapping = {
+            'Number': Number,
+        };
```
Comment (renestalder) [inline src/Resources/assets/controllers/input-mask_controller.js]: Nit: Just from an ergonomic perspective, this very static and can live outside the class.

File: src/Resources/assets/package.json [hunk 3]
```diff
@@ -15,6 +15,12 @@
                 "fetch": "eager",
                 "enabled": true
             },
+            "input-mask": {
+                "main": "controllers/input-mask_controller.js",
+                "webpackMode": "eager",
+                "fetch": "eager",
```
Comment (renestalder) [inline src/Resources/assets/package.json]: If we can't utilize lazy loading here for some reason, it's even more important the used library is loaded asynchronously and not synchronously.

**Review by renestalder — CHANGES_REQUESTED** [review]
@marcwieland95 We need to find a way to use lazy loading on those JavaScript libraries. It's not a modern approach, creating one big chunk, and the way I understand this here, this is doing exactly that.

File: src/Resources/assets/controllers/input-mask_controller.js (hunk 1 above)
Comment (mawiswiss) [reply src/Resources/assets/controllers/input-mask_controller.js]: @renestalder I'm not sure if `"fetch": "lazy",` is doing all the magic automatically. I adjusted that for a couple of controllers now.

If this isn't not completely solving the issue please open an issue in the Meta repo. This is out of scope

Comment (renestalder) [reply src/Resources/assets/controllers/input-mask_controller.js]: @marcwieland95

> I'm not sure if "fetch": "lazy", is doing all the magic automatically.

Well, I'm neither, but you could look it up, couldn't you? And check the network tab to see if the library is loaded lazily.

At least use `await import(...)` to lazy load the external library on demand inside `connect` if you can't figure out how to change the handling of Symfony here. That's the minimum we can and should do on frontend side.

> This is out of scope

Performance is the minimum we are going to check if we add an external library in addition to minimal security. These things are never out of scope. If we can't reason on the performance impact of adding a library, we are not doing it right.

Comment (mawiswiss) [reply src/Resources/assets/controllers/input-mask_controller.js]: @renestalder Yes, I just checked. I guess `lazy` doesn't include the full file. `imask` was still added to the `app.js` file and wasn't stripped out in a vendor file.

I included now the async import and this does the trick.

**Review by renestalder — APPROVED** [review]
@marcwieland95 Now that looks better. Wasn't able to test the solution but the code looks good to me.

**Pull request merged**